# DAPT 618
## Assignment #1 (Individual Assignment)
### Due by 28-JAN-2025

## *Tom Jones*

---

I have borrowed some exercise questions from Think Python book. I encourage everyone to read at least the first chapter of the book, which is available online here.

Please use Jupyter Notebook to show your work and provide answers in Jupyter Notebook.

1.	In a print statement, what happens if you leave out one of the parentheses, or both?

In [2]:
print('testing

SyntaxError: unterminated string literal (detected at line 1) (1273576526.py, line 1)

Cell In[2], line 1
    print('testing
          ^
SyntaxError: unterminated string literal (detected at line 1)

In [6]:
print 'testing'

SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)? (3368403465.py, line 1)

Cell In[6], line 1
    print 'testing'
    ^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?

In each attempt, Python finds the error and prevented the code from being fully executed.

2.	If you are trying to print a string, what happens if you leave out one of the quotation marks, or both?

In [4]:
print(testing')

SyntaxError: unterminated string literal (detected at line 1) (3628285504.py, line 1)

In [8]:
print(testing)

NameError: name 'testing' is not defined

In [ ]:
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[8], line 1
----> 1 print(testing)

NameError: name 'testing' is not defined

In both attempts, Python spotted the error and prevented the code from being fully executed.

3.	You can use a minus sign to make a negative number like -2. What happens if you put a plus sign before a number? What about 2++2?

In [10]:
+2 + 2

4

result: 4

In [12]:
2 ++ 2

4

result: 4 

In both cases, as 2 was already a positive integer, Python ignored the extra + sign and performed the proper calculation.

4.	In math notation, leading zeros are ok, as in 09. What happens if you try this in Python? What about 011?

In [14]:
09 + 1

SyntaxError: leading zeros in decimal integer literals are not permitted; use an 0o prefix for octal integers (2014528533.py, line 1)

Cell In[14], line 1
    09 + 1
    ^
SyntaxError: leading zeros in decimal integer literals are not permitted; use an 0o prefix for octal integers

In [16]:
011 + 1

SyntaxError: leading zeros in decimal integer literals are not permitted; use an 0o prefix for octal integers (4042067410.py, line 1)

  Cell In[16], line 1
    011 + 1
    ^
SyntaxError: leading zeros in decimal integer literals are not permitted; use an 0o prefix for octal integers

In both cases Python is believing this to be a case of bad syntax where it thinks I'm attempting to use octal formatting instead of regular integers, and thus throws an error preventing the code from being fully executed.

5.	How many miles are there in 10 kilometers? (Hint: There are 1.61 kilometers in a mile.)

In [18]:
km = 10
kmm = 1.61

miles = km / kmm
print(miles)

6.211180124223602

Though I could have told you 6.2 miles as a 10k is a standard running race distance of which I've ran few, the official result from this calculation is 6.211 (assuming some rounding is happening with the initial 1.61 km to 1 mile assumption and/or race course distances are being measured accordingly for the slight difference). 

6.	In Python, n = 42 is legal. What about 42 = n?

In [20]:
42 = n 

SyntaxError: cannot assign to literal here. Maybe you meant '==' instead of '='? (1187585252.py, line 1)

  Cell In[20], line 1
    42 = n
    ^
SyntaxError: cannot assign to literal here. Maybe you meant '==' instead of '='?

Python is throwing this error because although you can assign numeric values on the right side of the equal sign to a string/character variable name (n = 42), and you can assign one variable to another (p = n), numeric values themselves are not inherently considered variables thus cannot have values assigned to them. 

7.	How about x = y = 1?

In [28]:
x = y = 1

print(x, y)

1 1


result: 1 1
This proves what I stated above - you can assign a variable to another variable in addition to assigning a number to a variable.

8.	In some languages, every statement ends with a semi-colon. What happens if you put a semi-colon at the end of a Python statement?

In [36]:
x = y = 1

print(x, y);

1 1


result: 1 1
Regardless of whether I have one semicolon after the print statement, one semicolon after the variable assignment statement, or a semicolon after both, Python seems to ignore its existence and provide the correct output.

9.	What if you put a period at the end of a statement? What about an extra space at the end of a statement?

In [38]:
x = y = 1

print(x, y).

SyntaxError: invalid syntax (3385482583.py, line 3)

  Cell In[38], line 3
    print(x, y).
                ^
SyntaxError: invalid syntax

In [40]:
x = y = 1

print(x, y) 

1 1


result: 1 1

Extra spaces are ignored (though indentation ahead of characters at the start of new lines matters), periods are considered within the syntax rules as periods usually are used to access functions (example: list_variable.append(4)).

10.	Write a function named right_justify that takes a string named s as a parameter and prints the string with enough leading spaces so that the last letter of the string is in column 70 of the display.
```
>>> right_justify('monty')

                                                                 Monty
```

In [1]:
def right_justify(s):
    blank_space = 70 - len(s)
    ttl_space = ' ' * blank_space
    print(ttl_space + s)

right_justify('monty')   



                                                                 monty


11.	Write a function that draws a grid like the following:
```
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
```

In [76]:
def draw_grid():
    h_line = '+' + ' -'*4 + ' +' + ' -'*4 + ' +'
    v_line = '|' + ' '*9 + '|' + ' '*9 + '|'
    for i in range(2):
        print(h_line)
        for j in range(4):
            print(v_line)
    print(h_line)

draw_grid()
    

+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
